File: Data Structures & Algorithms/same-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // for a two tree to be identical ->
        // Are both NULL?
        // Are only one NULL?
        // Do their values differ?

        if(p == nullptr && q == nullptr){
            return true;
        }

        // if one is null and other is not null -> then not identical
        if(p == nullptr || q == nullptr){
            return false;
        }


        // if both values arent same , then false;
        if(p->val != q->val){
            return false;
        }

        // now check for thier left and right subtree and so on for evry node
        return isSameTree(p->left,q->left) && isSameTree(p->right , q->right);
    }
};

File: Data Structures & Algorithms/house-robber-iii/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int rob(TreeNode root) {
        // this is a very important hard tier oa question although it looks simple
        // its main logic are in notes please check
        // we will store mianly 2 values and find a max from one of them so , we start it and a helper function will privode it
        int[] masterchoices = fillBackPack(root);
        // this array will consist of only 2 values so we will return the max from one of them
        return Math.max(masterchoices[0] ,masterchoices[1]);
    }

    // now our very very imporant helper fucntion -> this is a recurive helper
    public int[] fillBackPack(TreeNode root){
        // since this is a recursion -> base case willl be 
        if(root == null){
            return new int[]{0,0};
        }

        // since we r following bottom to top appraoch -> we will use post order aprach so drill it to down
        int[] leftBackPack = fillBackPack(root.left);
        int[] rightBackPack = fillBackPack(root.right);

        // now step 2: after reading the downmost node we take 2 values from these arrays
        // these values rpresnts children of a node so
        int leftRobbed = leftBackPack[0];
        int leftSkipped = leftBackPack[1];

        int rightRobbed = rightBackPack[0];
        int rightSkipped = rightBackPack[1];

        // step 3 : very impormt imp part of this function
        // the main rule of this problem is that at each node we can have to choices
        // one is to robb this node and we should compulsoryly skip our child and go to its child
        // another is that if we r skiiping this node , we can have 2 possibilities ->either we can robb our childer node or we can skip our children node depending upon its value

        // choice a : 
        int robThisHouse = root.val + leftSkipped + rightSkipped;
        // choice B :
        int skipThisHouse = Math.max(leftSkipped , leftRobbed) + Math.max(rightSkipped , rightRobbed);

        // now we have calculated our choics , pass this choices to our parent nodes so they can use this
        return new int[]{robThisHouse , skipThisHouse};
          
    }
}